### PR TITLE
Integration changes working

### PR DESCRIPTION
**Pull request by MalekiRe**

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs [hunk 1]
```diff
@@ -162,6 +162,19 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
             return await check.Expression.EvalAsync(parameters, cancel);
         }
 
+        internal IEnumerable<ExpressionError> ParseAndRegisterUserDefinedFunctions(string script)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:165]: Is this just for tests?

Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs]: Changed

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs [hunk 2]
```diff
@@ -162,6 +162,19 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
             return await check.Expression.EvalAsync(parameters, cancel);
         }
 
+        public (IEnumerable<ExpressionError>, IEnumerable<FunctionInfo>) DefineFunctions(string script)
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:165]: Do you think we should wrap the members of this tuple up in a class with named properties?

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs [hunk 3]
```diff
@@ -162,6 +162,19 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
             return await check.Expression.EvalAsync(parameters, cancel);
         }
 
+        public (IEnumerable<ExpressionError>, IEnumerable<FunctionInfo>) DefineFunctions(string script)
+        {
+            var parsedUDFS = new Core.Syntax.ParsedUDFs(script);
+            var result = parsedUDFS.GetParsed();
+
+            var udfDefinitions = result.UDFs.Select(udf => new UDFDefinition(
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:170]: Are we checking to see if the actual return type matches the stated return type?

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/UDFLazyBinder.cs [hunk 4]
```diff
@@ -8,7 +8,7 @@ namespace Microsoft.PowerFx.Interpreter
 {
     internal class UDFLazyBinder
     {
-        private readonly UserDefinedTexlFunction _function;
+        public readonly UserDefinedTexlFunction _function;
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/UDF/UDFLazyBinder.cs:11]: Why make this field public?

**Review by bryancroteau-MSFT — CHANGES_REQUESTED** [review]

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/UDFLazyBinder.cs (hunk 4 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Interpreter/UDF/UDFLazyBinder.cs:11]: So we can read it to make the FunctionInfo

Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/UDF/UDFLazyBinder.cs:11]: Don't make this public. Instead, make a property

public UserDefinedFunction Function => _function;

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs (hunk 2 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:165]: +1 that would be cleaner. 
And the perk is we can also add new properties without it being a breaking change.

Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:165]: And maybe derive from IOperationStatus (like CheckResult does)

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs (hunk 3 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:170]: done

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/DefineFunctionsResult.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.PowerFx.Interpreter.UDF
+{
+    public class DefineFunctionsResult
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/UDF/DefineFunctionsResult.cs:10]: >DefineFunctionsResult [](http://example.com/codeflow?start=17&length=21)

Can this impl IOperationStatus, like CheckResult.

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs [hunk 6]
```diff
@@ -57,6 +57,15 @@ public IEnumerable<ExpressionError> Bind()
                 return errorList;
             }
 
+            if (!check.ReturnType._type.Kind.Equals(ReturnType.Kind))
+            {
+                var errorList = new List<ExpressionError>
+                {
+                    new ExpressionError()
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs:64]: >new ExpressionError() [](http://example.com/codeflow?start=20&length=21)

can these errors have a message?

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs [hunk 7]
```diff
@@ -162,6 +162,19 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
             return await check.Expression.EvalAsync(parameters, cancel);
         }
 
+        public DefineFunctionsResult DefineFunctions(string script)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:165]: >DefineFunctions [](http://example.com/codeflow?start=37&length=15)

Can this method be static?
Ie, is it actually changing anything on Engine? Or just exposing Parse?

Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:165]: This method cannot be static

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/DefineFunctionsResult.cs (hunk 5 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Interpreter/UDF/DefineFunctionsResult.cs]: Done

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs (hunk 6 above)
Comment (MalekiRe) [reply src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs:67]: Done

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/DefineFunctionsResult.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        
+        public DefineFunctionsResult(IEnumerable<ExpressionError> expressionErrors, IEnumerable<FunctionInfo> functionInfos)
+        {
+            _expressionErrors = expressionErrors;
+            _functionInfos = functionInfos;
+        }
+
+        public IEnumerable<ExpressionError> Errors => _expressionErrors;
+
+        public bool IsSuccess => _expressionErrors.Any();
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/UDF/DefineFunctionsResult.cs:26]: This should be inverted right? If Any returns true, IsSuccess is false.

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs [hunk 9]
```diff
@@ -53,6 +53,21 @@ public IEnumerable<ExpressionError> Bind()
...
+                return errorList;
+            }
+
+            if (!check.ReturnType._type.Kind.Equals(ReturnType.Kind))
+            {
+                var errorList = new List<ExpressionError>
+                {
+                    new ExpressionError()
+                    {
+                        Message = "Return type does not match body return type"
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs:69]: "The stated function return type does not match the return type of the function body"

File: src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs [hunk 10]
```diff
@@ -53,6 +53,21 @@ public IEnumerable<ExpressionError> Bind()
                 var errorList = new List<ExpressionError>
                 {
                     new ExpressionError()
+                    {
+                        Message = "Expression is not a ParsedExpression"
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/UDF/UserDefinedTexlFunction.cs:57]: Is this supposed to be an ExpressionError? Or an exception?

**Review by bryancroteau-MSFT — CHANGES_REQUESTED** [review]

**Review by bryancroteau-MSFT — APPROVED** [review]

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**